File: codes/Garnetwzy/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(root == NULL)
            return true;
        stack<TreeNode *> s;
        TreeNode *pre = NULL;
        while(!s.empty() || root) {
            while(root) {
                s.push(root);
                root = root->left;
            }
            if(!s.empty()) {
                root = s.top();
                s.pop();
                if(pre &&pre->val >= root->val) {
                    return false;
                }    
                pre = root;
                root = root->right;
            }
        }
        return true;
    }
};

File: codes/HellDream/98.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> list = new ArrayList<>();
        preOrder(list, root);
        for(int i=0;i<list.size()-1;i++){
            if(list.get(i)>=list.get(i+1)) return false;
        }
        return true;
    }
    public void preOrder(ArrayList<Integer> list,TreeNode root){
        if(root==null) return;
        if(root.left!=null) preOrder(list, root.left);
        list.add(root.val);
        if(root.right!=null) preOrder(list, root.right);
        
    }
}